sdk/servicebus/Azure.Messaging.ServiceBus/src/Diagnostics/ServiceBusEventSource.cs

```diff
@@ -1509,7 +1590,34 @@ public void TransactionDischarged(string localTransactionId, ArraySegment<byte>
         [Event(TransactionDischargedEvent, Level = EventLevel.Informational, Message = "AmqpTransactionDischarged for LocalTransactionId: {0} AmqpTransactionId: {1} Rollback: {2}.")]
         public void TransactionDischarged(string transactionId, string amqpTransactionId, bool rollback)
         {
-            WriteEvent(TransactionDischargedEvent, transactionId, amqpTransactionId, rollback);
+            if (IsEnabled())
+            {
+                TransactionDischargedCore(TransactionDischargedEvent, transactionId, amqpTransactionId, rollback);
+            }
+        }
+
+        [NonEvent]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private unsafe void TransactionDischargedCore(int eventId, string transactionId, string amqpTransactionId, bool rollback)
+        {
+            fixed (char* transactionIdPtr = transactionId)
+            fixed (char* amqpTransactionIdPtr = amqpTransactionId)
+            {
+                var eventPayload = stackalloc EventData[3];
+
+                eventPayload[0].Size = (transactionId.Length + 1) * sizeof(char);
+                eventPayload[0].DataPointer = (IntPtr)transactionIdPtr;
+
+                eventPayload[1].Size = (amqpTransactionId.Length + 1) * sizeof(char);
+                eventPayload[1].DataPointer = (IntPtr)amqpTransactionIdPtr;
+
+                // bool maps to "win:Boolean", a 4-byte boolean
+                var rollbackInt = Convert.ToInt32(rollback);
```

Comment:
There are other events that use bool parameter. Why do we only need to update this one?

Reply:
Because I was so excited that I found the root cause that I forgot to search for other occurrences of this problem. Will update. Thanks for your patience 😅

Reply:
Hehe no worries - I wasn't sure if we were missing coverage for that one, but I guess we didn't run the tests after your last commit. Let me try that now to verify that we have coverage.

Reply:
Probably a more efficient way for me to determine this would be to just look in our tests 😆 
It doesn't look like we have coverage for the LinkStateLost event source event. I think this one would be difficult to test as a live event source test, but hopefully we can add a mock test for it in EventSourceTests.

Reply:
It uses a test event listener which is similar to my repro code. It would have caught it. I can see what I can do at some later point

Reply:
I briefly looked at the EventSourceTests and since they use the higher level API surface to simulate the client into the cases that triggers the event log entry it would be quite involved for me to understand how to mock the client into a situation where we get a linked lost within the right state. So if any of you can quickly bang this out I wouldn't mind direct pushes to the branch. Otherwise it is going to take a while for me to get to this.

Reply:
Appreciate your patience @danielmarbach - I'll try to look into adding a test sometime this week.

Reply:
I had a quick look, started adding an internal property to the `AmqpReceiver` that would make it possible to override the event source. Then I ran into the rabbit hole of somehow having to mock away the `ReceivingAmqpLink` which I couldn't find out with reasonable effort at this stage. 

Then I also realized we would need to figure out where to place the tests because the `EventSourceTests` currently test from the higher level while the event log in question was for the lower level AMQP connection events which seems we haven't really tried to unit test. That in itself is probably a lot of effort for me to get working. Given there is no real test coverage of these paths I'm inclined to say that improvement could be done separately from this PR (but of course I'm biased to get this in :) )

Reply:
So I tried closing the underlying receive link but that ends up resulting in a lock lost exception rather than the OperationCanceledException. Because we can't mock the underlying AMQP methods it seems like it will be hard to reliably test this. I think we can go ahead with this as is, since as you mentioned, we didn't have coverage for this event before either.